### PR TITLE
Enrichment and redaction

### LINKED ISSUE #31844: HttpLogging redaction and enrichment
We need the ability to allow users to configure and customize what is logged for a request and response. For example if they would like to modify a header value that is logged to remove PII, or adding another key and value to the log.

Our current idea is to have a Func that people can implement that would take a context, and allow returning any KeyValuePair to be logged.

Currently, we were thinking of adding the extension point to where we log headers:

General API shape
```c#

Action<HttpLoggingContext> ModifyLog;

HttpLogContext
{
    HttpContext HttpContext { get; }
    List<KeyValuePair<string, string>> Logs { get; } // or maybe a dictionary
}
```

Example:

```C#
(httpLogContext) =>
{
   httpLogContext.Logs.Add(
       new KeyValuePair<string, string>(
           "RequestAborted", 
            httpLogContext.HttpContext.RequestAborted.IsCancellationRequested.ToString());
}
```

### PR DESCRIPTION
**Pull request by jkotalik**
Fixes https://github.com/dotnet/aspnetcore/issues/31844.

Current iteration of the API design. Would like feedback on how this feels before undrafting (there are many perf improvements we can make, this is just getting it "working").

### CODE REVIEW

**Comment by jkotalik** [comment]
cc @Tratcher @halter73 @davidfowl @shirhatti if you have opinions 😄

File: src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs [hunk 1]
```diff
@@ -82,35 +83,114 @@ private async Task InvokeInternal(HttpContext context)
...
-                    AddToList(list, nameof(request.Protocol), request.Protocol);
-                }
+                    var headerDictionary = new HeaderDictionary();
+                    var loggingContext = new HttpRequestLoggingContext(context, options, headerDictionary);
 
-                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestMethod))
-                {
-                    AddToList(list, nameof(request.Method), request.Method);
-                }
+                    if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestProtocol))
```
Comment (Tratcher) [inline src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs:91]: InvokeInternal is too big. Extract these `if` groups to helper methods so we only see the main control flow here.

File: src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs [hunk 2]
```diff
@@ -82,35 +83,114 @@ private async Task InvokeInternal(HttpContext context)
...
+                    if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestPath))
+                    {
+                        loggingContext.PathBase = request.PathBase;
+                        loggingContext.Path = request.Path;
+                    }
 
-                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestHeaders))
+                    if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestQuery))
+                    {
+                        loggingContext.Query = request.QueryString.ToString();
```
Comment (Tratcher) [inline src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs:114]: Using PathString.Value and QueryString.Value would avoid extra re-encoding, unless we know people would prefer the encoded value in the logs?

**Review by Tratcher — COMMENTED** [review]
> if you have opinions 😄

Of course we do, just not about anything important.

> (there are many perf improvements we can make, this is just getting it "working").

👍 Those can wait for another PR, there's already a lot going on here.

File: src/Middleware/HttpLogging/src/HttpRequestLoggingContext.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        public List<(string, string)> Extra
+        {
+            get
+            {
+                _extra ??= new List<(string, string)>();
+                return _extra;
+            }
+        }
+    }
+}
```
Comment (davidfowl) [inline src/Middleware/HttpLogging/src/HttpRequestLoggingContext.cs:81]:
```suggestion
        public List<(string, string)> Extra => _extra ??= new List<(string, string)>();
```

File: src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs (hunk 2 above)
Comment (shirhatti) [reply src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs:114]: I would have assumed people prefer the unencoded value in the logs.

**Comment by BrennanConroy** [comment]
What's the status of this change?

**Review by dougbu — CHANGES_REQUESTED** [review]
New files need MIT license headers

**Comment by jkotalik** [comment]
@wtgodbe I'm going to close this for now as it's been open for a while, feel free to pick up if necessary.

**Pull request closed without merging**